mail_embed_image/models/ir_mail_server.py

```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 Therp BV <https://therp.nl>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+import requests
+import uuid
+from odoo import models
+from odoo.addons.base.ir.ir_mail_server import encode_header_param
+from lxml.html.soupparser import fromstring
+from lxml import etree
+from base64 import b64encode
+from email.mime.image import MIMEImage
+from email import Encoders
+
+
+class IrMailServer(models.Model):
+    _inherit = 'ir.mail_server'
+
+    def build_email(
+            self,
+            email_from,
+            email_to,
+            subject,
+            body,
+            email_cc=None,
+            email_bcc=None,
+            reply_to=None,
+            attachments=None,
+            message_id=None,
+            references=None,
+            object_id=None,
+            subtype='plain',
+            headers=None,
+            body_alternative=None,
+            subtype_alternative='plain'):
+        result = super(IrMailServer, self).build_email(
+            email_from,
+            email_to,
+            subject,
+            body,
+            email_cc=email_cc,
+            email_bcc=email_bcc,
+            reply_to=reply_to,
+            attachments=attachments,
+            message_id=message_id,
+            references=references,
+            object_id=object_id,
+            subtype=subtype,
+            headers=headers,
+            body_alternative=body_alternative,
+            subtype_alternative=subtype_alternative,
+        )
+
+        def _build_email_replace_img_src(msg, attachments):
+            """ Given a message, find it's img tags and if they
+            are URLs, replace them with cids.
+
+            :param msg: A email.message.Message
+            """
+            if msg.is_multipart():
+                for part in msg.get_payload():
+                    _build_email_replace_img_src(part, attachments)
+            else:
+                if msg.get_content_subtype() == 'html':
+                    body = msg.get_payload(decode=True)
+                    root = fromstring(body)
+                    for img in root.xpath(
+                            "//img[starts-with(@src, '/web/image')]"):
+                        base_url = self.env['ir.config_parameter'].get_param(
+                            'web.base.url')
+                        src = base_url + img.get('src')
+                        response = requests.get(src)
```

Comment:
as you now enforce that this is an image linked from Odoo, you can just as well determine which attachment this is from the address and read this via the ORM, less overhead

Reply:
I am not sure how you want me to do this.

At this point I do know that the url for an image stored in the filestore begins with `/web/image/` after that a lot of things can exist - including the url containing dimensions of the image that when given the image is resized. All this happens in the controller, so doesn't it make sense to just have the controller take care of everything? Let alone that that request might be forwarded to a different server in different configurations.

FYI at this point all outlook, gmail and Therp's MUAs show the email correctly

Reply:
look at the [controller for images](https://github.com/OCA/OCB/edit/10.0/addons/web/controllers/main.py#L1007) and then at the definition of `binary_content`

Great the PoC works, now you can deploy this for the customer to test with their version of outlook, add proper tests and open the PR here